libraries/AP_HAL_ChibiOS/AnalogIn.cpp

```diff
@@ -206,233 +236,389 @@ void AnalogSource::_add_value(float v, float vcc5V)
+            adcgrpcfg[index].sqr[0] |= chan << (6*(i+1));
         } else if (i < 9) {
-            adcgrpcfg.sqr[1] |= chan << (6*(i-4));
+            adcgrpcfg[index].sqr[1] |= chan << (6*(i-4));
         } else {
-            adcgrpcfg.sqr[2] |= chan << (6*(i-9));
+            adcgrpcfg[index].sqr[2] |= chan << (6*(i-9));
         }
 #else
         if (chan < 10) {
-            adcgrpcfg.smpr2 |= ADC_SAMPLE_480 << (3*chan);
+            adcgrpcfg[index].smpr2 |= ADC_SAMPLE_480 << (3*chan);
         } else {
-            adcgrpcfg.smpr1 |= ADC_SAMPLE_480 << (3*(chan-10));
+            adcgrpcfg[index].smpr1 |= ADC_SAMPLE_480 << (3*(chan-10));
         }
         // setup channel sequence
         if (i < 6) {
-            adcgrpcfg.sqr3 |= chan << (5*i);
+            adcgrpcfg[index].sqr3 |= chan << (5*i);
         } else if (i < 12) {
-            adcgrpcfg.sqr2 |= chan << (5*(i-6));
+            adcgrpcfg[index].sqr2 |= chan << (5*(i-6));
         } else {
-            adcgrpcfg.sqr1 |= chan << (5*(i-12));
+            adcgrpcfg[index].sqr1 |= chan << (5*(i-12));
         }
 #endif
     }
-    adcStartConversion(&ADCD1, &adcgrpcfg, samples, ADC_DMA_BUF_DEPTH);
-
-#if HAL_WITH_MCU_MONITORING
-    setup_adc3();
+#if STM32_ADC_DUAL_MODE
+    // assert that ADC1 and ADC2 have the same number of channels
+    static_assert(ARRAY_SIZE(AnalogIn::pin_config) == ARRAY_SIZE(AnalogIn::pin_config_2), "ADC1 and ADC2 must have same num of channels");
+
+    if (index == 0) {
+        for (uint8_t i=0; i<num_grp_channels; i++) {
+            uint8_t chan = get_pin_channel(1, i);
+            // setup cycles per sample for the channel
+            adcgrpcfg[0].pcsel |= (1<<chan);
+            adcgrpcfg[0].ssmpr[chan/10] |= ADC_SMPR_SMP_384P5 << (3*(chan%10));
+            if (i < 4) {
+                adcgrpcfg[0].ssqr[0] |= chan << (6*(i+1));
+            } else if (i < 9) {
+                adcgrpcfg[0].ssqr[1] |= chan << (6*(i-4));
+            } else {
+                adcgrpcfg[0].ssqr[2] |= chan << (6*(i-9));
+            }
+        }
+    }
 #endif
+
+    adcStartConversion(adcp, &adcgrpcfg[index], samples[index], ADC_DMA_BUF_DEPTH);
+    return;
+failed_alloc:
+    AP_HAL::panic("Failed to allocate ADC DMA buffer");
 }
 
 /*
   calculate average sample since last read for all channels
  */
-void AnalogIn::read_adc(uint32_t *val)
+void AnalogIn::read_adc(uint8_t index, uint32_t *val)
 {
     chSysLock();
-    for (uint8_t i = 0; i < ADC_GRP1_NUM_CHANNELS; i++) {
-        val[i] = sample_sum[i] / sample_count;
+    uint8_t num_grp_channels = get_num_grp_channels(index);
+    if (num_grp_channels == 0) {
+        chSysUnlock();
+        return;
     }
-    memset(sample_sum, 0, sizeof(sample_sum));
-    sample_count = 0;
-    chSysUnlock();
-}
-
-
+    for (uint8_t i = 0; i < num_grp_channels; i++) {
+        val[i] = sample_sum[index][i] / sample_count[index];
+    }
+    memset(sample_sum[index], 0, sizeof(uint32_t) * num_grp_channels);
```

Comment:
adcsample_t?

Reply:
In Dual mode adcsample_t is 32bit. I felt this is clearer.